### ISSUE TITLE
download-from-binary-cache.pl: still waiting for ‘http://cache.nixos.org/f7pxpq9f1cmb934f9v3a14qyhq3h54ry.narinfo’ after 5 seconds...

### ISSUE THREAD

**Issue by wwat** [NixOS/nixpkgs#5546]
```
download-from-binary-cache.pl: still waiting for ‘http://cache.nixos.org f7pxpq9f1cmb934f9v3a14qyhq3h54ry.narinfo’ after 5 seconds...
```
I can use `links` just fine to browse `http://nixos.org` from that machine. 
```
# 30 minutes ago nix-env -i firefox worked fine, but now I also get the below:
nix-env -i ghc
installing `ghc-7.8.3-wrapper'
download-from-binary-cache.pl: still waiting for ‘http://cache.nixos.org/ncbjrdbg3mrl1rj9a0d0aaz55sz3kdk1.narinfo’ after 5 seconds...
```
When I copied a random URL containing cache.nixos.org, I was able to access it from another machine (on the same network). Either way, weirdness.

**Comment by wwat** [comment]
3 minutes later all of a sudden the ghc command started to download something and completed successfully, but the `nixos-rebuild switch --upgrade` command doesn't seem to make that same progress.

**Comment by wwat** [comment]
I reset my ISP provided router and then it started working.

**Comment by wwat** [comment]
I am now fairly sure that this is just a major bug in nixos-rebuild, because when I killed it (when it was waiting again basically forever (>30 seconds)) and then ran `nixos-rebuild switch` again, everything magically started to work again.

**Comment by vcunat** [comment]
I don't think so. The binary cache is a little lazy sometimes, probably when under load or when Amazon servers are under load. For me the answers can commonly take half a minute, IIRC.

**Comment by iElectric** [comment]
It's just taking time to get the full response back.

**Comment by wwat** [comment]
Well, then it is a user interface problem. The way it should work is that it first establishes that the network connection of the user is not dead (download google.com), then that cache.nixos.org is not in fact dead, and then finally it should give some finite number as to how long it is going to take approximately before program failure (as opposed to some other source of failure) can be reported by a user. I noticed that it opened close to 1000 connections, and for consumer routers, the limit is around that, so that probably also doesn't help.

**Comment by iElectric** [comment]
It's been long known that it opens lots of connections and some firewalls will block such behavior. I'm not sure if someone looked into it to see if there is some space for optimization(s).

**Comment by vcunat** [comment]
The messages
```
download-from-binary-cache.pl: still waiting for ‘http://cache.nixos.org f7pxpq9f1cmb934f9v3a14qyhq3h54ry.narinfo’ after 5 seconds...
```
are no failures! They're just notifications that it's waiting on the server, so the user doesn't think it just hangs. If the an URL is unreachable, it fails immediately (because the connection fails fast or returns 404).

**Comment by wwat** [comment]
If there is no progress (bandwidth being used was close to 0), then it just means it is broken. In the end every program which is too slow under real world conditions is a broken program. If NixOS would take 10 days to install, nobody would use it. In a certain sense it is thus a real-time program. The real-time property to achieve here is that the user connection should be saturated or the upstream server should be saturated within a reasonable amount of time or it should be growing visibly to that state.

**Comment by vcunat** [comment]
:-D The only solution I see is to spend money on better servers for those files.

**Comment by iElectric** [comment]
Those files are served via S3, so that's not a problem.

**Comment by wwat** [comment]
You do know that S3 has been *designed* to have imperfections, right?

**Comment by iElectric** [comment]
@wwat: does running it with `-j 1` help?

**Comment by vandenoever** [comment]
I've found that using -j 1 with nixos-rebuild is a workaround for this problem.

**Comment by iElectric** [comment]
I wonder if you're hitting the file limit. Grab pid of nix-daemon and check `/proc/<pid>/fd/` and `/proc/<pid>/limits`

**Comment by edolstra** [comment]
Generally, if getting .narinfo files from cache.nixos.org seems to hang, it's because of (broken) NAT routers that don't like the fact that download-from-binary-cache uses a high number of parallel TCP connections. If that's the case, you can put something like "binary-caches-parallel-connections = 10" in nix.conf to reduce the amount of parallelism. (The default is 150 connections.)

Hopefully, one day S3/CloudFront will support SPDY/HTTP2 and we can use a single TCP connection.

**Comment by edolstra** [comment]
BTW, this is not a Nixpkgs but a Nix issue.

**Comment by vcunat** [comment]
I think we can close this.

**Comment by vandenoever** [comment]
While upgrading machine to NixOS 14.12, I've come across the 'still waiting' error again.
I've tried using the -j 1 option to nixos-rebuild and also adding this to /etc/nixos/configuration.nix:
  nix.extraOptions = ''
    binary-caches-parallel-connections = 1
  '';
Even with both options, the program nixos-rebuild still hangs with tens of 'still waiting' messages. The binary-caches-parallel-connections option seems to have no effect: if it did, then the # of messages would be lower. I also edited /etc/nix/nix.conf directly. With these 3 measures in place, I still get tens of 'still waiting' messages.
When I kill nixos-rebuild (ctrl-c) the process for download-from-binary-cache.pl stays alive and still has all the open IPv4 sockets.

After manually killing all the download-from-binary-cache.pl and using the 3 measures mentioned, I could run nixos-rebuild.

**Comment by danmilon** [comment]
Indeed, for me the issue was caused by the firewall of my ISP's router. I disabled it and it worked.

**Comment by wwat** [comment]
Why do you close this ticket when there is no solution?

**Comment by lethalman** [comment]
It happens sometimes to me too, even if I don't have any network issue. And I know it happens to other people too. It's not known whether it's a program issue, or server issue yet. Reopening.

**Comment by vcunat** [comment]
No, I closed this intentionally, as suggested by my comment https://github.com/NixOS/nixpkgs/issues/5546#issuecomment-68712734.

I don't think there is a problem in nix(os) itself, but rather some people's networks not liking so many parallel connections. Maybe I'm wrong...

**Comment by lethalman** [comment]
@vcunat my connectivity is perfectly fine, it's either a problem of nixos server or the program itself downloading from the binary cache, or the perl library or whatelse.

**Comment by danmilon** [comment]
@lethalman for me the issue was my ISPs router that has some pre-enabled firewall rules was blocking these specific connections. I did not inspect further since disabling the firewall fixed the issue.

**Comment by wwat** [comment]
The goal is to be able to download packages; a program is a method of obtaining those. If the program fails, it's a broken program. If in fact there is some limitation in the firewall, the program could actually run a self-test to check the limitations of the network. That the environment is slightly more challenging than you wish for has nothing to do with it.  It just requires a more intelligent program to solve rather than the naive attempt which has currently been done.

**Comment by vcunat** [comment]
Sure, there may be multiple ways to experience this.

**Comment by aragnon** [comment]
```
nix-env --upgrade
error: substituter ‘download-from-binary-cache.pl’ died unexpectedly
```
```nix-env --version```
```nix-env (Nix) 1.8```

Clearly this program needs work.

**Comment by pesterhazy** [comment]
I'm still seeing this problem intermittently on both nixos (using `nixos-rebuild`) and darwin (using `nix-env -i`).

Passing `-j 1` had no effect for me. What worked for me: With `nix-env` you can pass `--option use-binary-caches false` to disable binary caches temporarily. Clearly not a proper solution, because you have to build everything from source.

**Comment by aragnon** [comment]
@pesterhazy Is that on the same network? For me, it works, if I just try 3 times in a row on some networks.

**Comment by pesterhazy** [comment]
@aragon, I've encountered the issue both on my home DSL network and on a coffee shop. Both have adequate internet connections and I didn't experience problems with other services at the same time. I still cannot rule out that it's network related of course.

Retrying doesn't help for me.

**Comment by aragnon** [comment]
@pesterhazy Home DSL or coffee shop doesn't really sound very high-grade. Just to be clear, I think the program is sub-optimal for apparently its current users, but proper networking hardware does actually solve the problem. (On a network where I do have proper connectivity, I have no issues. )

A simple solution would be to eliminate all sources of parallelism from the program to at least ensure progress, so if you build nix from source, that should be fairly easy to do. Complaining to your ISP generally does not work (unfortunately). Did you already try all of the suggestions in this thread too?

**Comment by pesterhazy** [comment]
@aragon It's hard to test because of the intermittent nature of the problem. Right now everything seems to be working again, perhaps because the local cache is now filled.

I've tried `-j` and setting an option in nix.conf, neither of which seemed to have an effect.

I'll report back the next time I see the error.

**Comment by DonManley** [comment]
I just encountered the same problem. Passing `--option use-binary-caches false` worked, while `-j 1` did nothing. 

That being said, it still fails, but at least for another reason ("cannot download nghttp2-0.7.14.tar.bz2 from any mirror"). I was trying to `nix-env -iA nixpkgs.leksah` or `nix-env -i leksah".

**Comment by vcunat** [comment]
That nghttp2 problem is another thing. @wkennington: your http://pub.wak.io/ doesn't work anymore.

**Comment by obadz** [comment]
I have the same issue with a load balanced connection between a 15Mbit/sec DSL and a 50Mbit/s 4G connexion (the latter is CGNATed).

binary-caches-parallel-connections = 1 still prints tons of these warnings
-j 1 doesn't fix it either

**Comment by MarkusBarthlen** [comment]
@danmilon mentioned that the firewall of the router might need to be adjusted. Are there any specific ports that I would need to open? Thank you

**Comment by danse** [comment]
I hit this today, the solution proposed by @edolstra allowed me to overcome the problem, so i think that the problem was caused by the 150 parallel connections, which my Internet Service Provider, or somebody in between, did not seem to like.

A couple of remarks:
- it has been suggested here to use `-j 1`. according to the documentation this limits the amount of parallel jobs to 1, which is the default anyway. i am not sure that this helps at any extent reducing the number of parallel connections, or at least for many users it won't
- i tried initially to set the number of parallel connection to 1 with `$ nix-env --option binary-caches-parallel-connections 1 -i ...`, but i could see many parallel connections still failing, so probably this parameter is too low and it gets ignored. i had success setting it to 10 with `$ nix-env --option binary-caches-parallel-connections 10 -i ...`

**Comment by vcunat** [comment]
Yes, I'm fairly sure "-j" only takes effect *after* all "http://*.narinfo" are downloaded. @MarkusBarthlen: note these are plain HTTP, no special ports.

@danse: the code doesn't seem to handle any number above zero specially.
